Add container build workflow, and Dependabot

shymega opened the pull request:
This commit adds support for Dependabot, and a container build workflow, that, upon pushes to the repo, but *not* pull requests, builds a container image at one at a time, and pushes to GitHub Registry.

Signed-off-by: Dom Rodriguez <shymega@shymega.org.uk>

shymega commented:
This PR could be split to add support for Dependabot and the container workflow into two separate PRs - let me know. Thanks.

ThatcherC commented:
This is cool! Do you think it's worth adding an additional stage to the docker file so that the `apt installs` and the GCC installation are docker in one stage, and the `COPY . /usr/src` is done in another? That way the action wouldn't have to download the compile and apt stuff every time the action runs.

What I've had actions do in the past is:
- try to pull the base image (the one with the GCC and apt installs done) from a registry (like ghcr.io)
- if the image doesn't exist, build it and push to ie ghcr.io
- from the base image build a final image with the `COPY . /usr/src` step

Every time the action is run after the first time, only the ghcr pull and the `COPY . /usr/src` step need to be run, which is much faster than rebuilding the whole image every time.

I think this approach makes sense if we don't expect the Dockerfile to change very often. It also means we'll put less pressure on https://armkeil.blob.core.windows.net and have faster build times.

shymega commented:
@ThatcherC Thanks. I've updated it a bit, but it's difficult to combine into one image. We could have two container builds running, but that'd happen every push. How does my latest push look to you? And should I split out the Dependabot configuration to a new PR?

What would be cool is to use this PR's Dockerfile with #6.

shymega commented:
Actually, #6 already uses the Dockerfile in the parent directory - so ignore my confusion. When/if this PR is merged, #6 will use the Dockerfile in this PR.

shymega commented:
I've amended the last commit - the GCC path I changed was hardcoded in the build system, and it broke builds. I'm hoping this change fixes it.

shymega commented:
Just to check, should I split the Dependabot config out into a separate PR? The fw build seems to work, IIRC.

ThatcherC commented:
Ultimately a question for @Ralim I suppose, but seems like you might as well split them into separate PRs. That way if Dependabot could be merged even if the Dockerfile changes aren't accepted (or vice versa).

shymega commented:
OK, so what I've done is split my PR into four different branches on my fork. There's the Dockerfile refactor, container workflow, Dependabot, and finally, the Devcontainers PR which is still active. I will now make PRs for the three remaining branches.
